File: Portal.Model/domainclasses.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Web;

namespace Portal.Model
{
	public class Product
	{
		public int id { get; set; }
		public string code { get; set; }
		public string name { get; set; }
		public string description { get; set; }		
	}

	public class Customer
	{
		[Key]
		public string code { get; set; }
		public string name { get; set; }
		public string address1 { get; set; }
		public string address2 { get; set; }
		public string address3 { get; set; }
		public string address4 { get; set; }
		public string address5 { get; set; }
		public string address6 { get; set; }
		public string town_city { get; set; }
		public string county { get; set; }
		public string currency { get; set; }
		public string analysis_codes_1 { get; set; }
		public string invoice_customer { get; set; }
	}

	public class User
	{
		public int id { get; set; }
		public string name { get; set; }
		public string lastname { get; set; }
		public string username { get; set; }
		public string password { get; set; }
		public string email { get; set; }
		public string customer_code { get; set; }
		public string address { get; set; }
		public string phone { get; set; }
		public bool? isInternal { get; set; }
		public DateTime? lastLogin { get; set; }

		[NotMapped]
		public string token { get; set; }

		public virtual Customer Customer { get; set; }
		public virtual List<Role> Roles { get; set; }
		public virtual List<UserSession> Sessions { get; set; }

		[NotMapped]
		public List<Permission> Permissions { get; set; }

		public bool isAdmin => Roles?.Count(r => r.id == Role.Admin) > 0;
		public bool isBranchAdmin => Roles?.Count(r => r.id == Role.BranchAdmin) > 0;

		public bool HasPermission(PermissionId permId)
		{
			return Permissions?.Count(p => p.id == (int) permId) > 0;
		}
	}

	public class UserSession
	{
		public int id { get; set; }
		public int? user_id { get; set; }
		public DateTime? dateCreated { get; set; }
		public string token { get; set; }
		public string ip_addr { get; set; }
	}

	public class Role
	{
		public const int User = 1;
		public const int BranchAdmin = 2;
		public const int Admin = 3;

		public int id { get; set; }
		public string name { get; set; }

		public List<Permission> Permissions { get; set; }
	}
    public class Holiday
    {
        public int id { get; set; }
        public string name { get; set; }
        public DateTime date { get; set; }
    }

	[Table("permission")]
	public class Permission
	{
		public int id { get; set; }
		public string name { get; set; }
	}

	[Table("user_permission")]
	public class UserPermission
	{
		public int user_id { get; set; }
		public int permission_id { get; set; }
		public bool? grant { get; set; }
		public bool? deny { get; set; }

	}

	[Table("customer_permission")]
	public class CustomerPermission
	{
		public string customer_code { get; set; }
		public int permission_id { get; set; }
		public bool? grant { get; set; }
		public bool? deny { get; set; }

	}

	public enum PermissionId
	{
		ViewAccountDetails = 1,
		ViewHolidayAdministration = 2,
		ViewInvoiceHistory = 3,
		ViewOrderHistory = 4,
		ViewStockSearch = 5,
		ViewUserAdministration = 6,
		ViewRoleAdministration = 7
	}
}

File: Portal/angularapp/app/domainclasses.ts

export class Product {
    id: number;
    code: string;
    name: string;
    description: string;
    combined_name: string;
}

export class Customer {
    code: string;
    name: string;
    address1: string;
    address2: string;
    address3: string;
    address4: string;
    address5: string;
    address6: string;
    town_city: string;
    county: string;
    analysis_codes1: string;
    currency: string;
    combined_name: string;
}

export class User {
    id: number;
    name: string;
    lastname: string;
    username: string;
    password: string;
    email: string;
    customer_code: string;
    address: string;
    phone: string;
    token: string;
    customer: Customer;
    roles: Role[];
    locked: boolean;
    isTopAdmin: boolean;
    isBranchAdmin: boolean;
    activated: boolean;
    isInternal: boolean;
    lastLogin: Date;
    role: string;
    role_id: number;
    permissions: Permission[];

    constructor() {
        this.customer = new Customer();
    }
}

export class Role {
    id: number;
    name: string;
    permissions: Permission[];
    // strPermissions: string[];

    constructor(id: number, name?: string, permissions?: Permission[]) {
        this.id = id;
        this.name = name;
        this.permissions = permissions;
    }
}


export class Invoice {
    customer: string;
    dated: Date | null;
    kind: string;
    reference: string;
    item_no: string;
    due_date: Date | null;
    amount: number | null;
    open_indicator: string;
    customer_order_num: string;
    selected: boolean;
    postcode: string;
}

export enum RoleId {
    User = 1,
    BranchAdmin = 2,
    Admin = 3
}

export class Order {
    order_no: string; // Sage order number
    customer_order_no: string; // Customer Order number
    date_entered: Date; // Date order received and input to Sage
    invoice_date: Date; // Currently will be to indicate "Delivered" flag. When there is an invoice date then the order is delivered
    date_required: Date;
    address1: string;
    address2: string;
    address3: string;
    address4: string;
    address5: string;
    address6: string;
    town_city: string;
    statusText: string;
    status: number;
    customer?: string;
    country: string;
    audits: OrderAudit[];
    productStockData: StockData[];
    details: OrderDetail[];
    planned: boolean;
}

export class OrderDetail {
    product: string;
    order_qty: number;
    val: number;
    order_line_no: number;
}

export  enum OrderStatus {
    Received = 5,
    InProgress = 7,
    Delivered = 8,
    Cancelled = 9
    // Processed = 'Processed',
    // Confirmed = 'Confirmed',
    // Delivered = 'Delivered',
    // Cancelled = 'Cancelled',
}
export class OrderAudit {
    audit_key: string; // need to join to the Order_no field of opheadm
    character_value: string; // limited to 6 to infer Despatch Note printed
    audit_date: Date; // is the date used for Order Processed date.
}

export class StockData {
    quantity: number;
    product: Product;
    ship_date: Date | null;
    discontinued: boolean;
    madeToOrder: boolean;
    expiresThisYear: boolean;
}

export class Holiday {
    name: string;
    date: Date;
}
export class HolidayExtracted {
    name: string;
    year: number;
    month: number;
    date: number;
    fullDate: Date;
}
export class Permission {
    id: number;
    name: string;
    selected: boolean;

    constructor(id?: number, name?: string, selected?: boolean) {
        this.id = id;
        this.name = name;
        this.selected = selected;
    }
}

export enum PermissionId {
    ViewAccountDetails = 1,
    ViewHolidayAdministration = 2,
    ViewInvoiceHistory = 3,
    ViewOrderHistory = 4,
    ViewStockSearch = 5,
    ViewUserAdministration = 6,
    ViewRoleAdministration = 7
}

export class ClearanceStock {
    product_code: string;
    long_description: string;
    range: string;
    exvat_price: number | null;
    price: number | null;
    freestock: number | null;
    discount: number;
}
